File: gifts.gl2/fgl_tutor/src/workshop/double3.f
C  double3.f  --  This program uses double buffering to draw a
C		   dynamic image of a ball traveling over a field
C		   without flashing.  The motion is accomplished by
C		   redrawing the scene to the back buffer.  When an
C		   image is completed, the entire buffer is swapped
C		   to the fore, and its image is displayed.
C  calls introduced:  double, gconfi, swapbu,
C		       frontb

C  WARNING:  For machines with a minimum configuration of bitplanes,
C  there may not be enough bitplanes to run all the colors in this
C  and subsequent double buffered programs.  A minimum of 12
C  bitplanes are required for this program.  For subsequent programs,
C  a minimum of 16 bitplanes will be required.

#include "fgl.h"
#include "fdevice.h"

C  In the window manager, double buffered programs MUST swapbuffers,
C  even when idling.  The window manager waits for all active double
C  double buffered programs to issue a swapbuffers() call before
C  actually swapping buffers.  If a program idles without issuing
C  a swapbuffers() call, all other double buffered programs will
C  wait for it.	

       CALL INITIALIZE
       CALL DRAWIMAGE

 100   IF (QTEST() .NE. 0) THEN
            CALL PROCESSINPUT
       ENDIF
       GOTO 100

       END

C  To initiate double buffering, the doublebuffer() call is made.
C  The display mode change to double buffering does not take place
C  until the gconfig() call is made.

       SUBROUTINE INITIALIZE

#include "fgl.h"
#include "fdevice.h"

       INTEGER DUMMY

       CALL PREFSI(450, 450)
       DUMMY = WINOPE ('DIAMOND', 7)

C   display mode to become double buffer
       CALL DOUBLE
C   display mode change takes effect
       CALL GCONFI

       CALL QDEVIC( REDRAW )

C   make color of fences, lines
       CALL MAPCOL (8, 240, 240, 240)

C   make color of grass
       CALL MAPCOL (9, 0, 175, 0)

C   make color of ball
       CALL MAPCOL (10, 240, 150, 0)

       RETURN
       END

C  Process input from the window manager

       SUBROUTINE PROCESSINPUT

#include "fgl.h"
#include "fdevice.h"

       INTEGER*2 VAL
       INTEGER DEV

 300   IF (QTEST() .EQ. 0) GOTO 350
            DEV = QREAD(VAL)
	    IF (DEV .EQ. REDRAW) THEN
		CALL RESHAP
		CALL DRAWIMAGE
            ENDIF
            GOTO 300
 350   CONTINUE

       RETURN
       END

C  drawimage() creates motion which doesn't flash.  The for loop
C  changes the value of i, which in turn, changes the position of
C  the ball.  Each iteration through the loop, the field is redrawn
C  and the ball is redrawn in a new position.

       SUBROUTINE DRAWIMAGE

#include "fgl.h"
#include "fdevice.h"

       INTEGER I

       DO 400 I=0, 299, 3
            CALL COLOR(0)
            CALL CLEAR
            CALL DIAMOND

C   draw ball with color 10 at position (x, y)

            CALL COLOR(10)
            CALL CIRCFI(I, I, 3)

C   swapbuffers
            CALL SWAPBU

 400   CONTINUE

C  At the end of the loop, the ball will stop moving.  However,
C  the two buffers will have the ball in different positions. 
C  You want to be certain that the ball is in the same
C  position in both buffers.  Otherwise, when at rest, the
C  program will swap between the two buffers, and the ball
C  will appear to bounce back and forth between the disparate
C  positions in the two buffers.  To draw the same scene
C  into both buffers, use frontbuffer(TRUE).  Then draw	
C  the scene--that draws it into both front and back buffers.
C  frontbuffer(FALSE) restores the default state, where only
C  the back buffer is drawn into.

       CALL FRONTB(.TRUE.)
       CALL COLOR(0)
       CALL CLEAR
       CALL DIAMOND
       CALL COLOR(10)
       CALL CIRCFI(I, I, 3)
       CALL FRONTB(.FALSE.)

       RETURN
       END

C   diamond has not changed since last example.

       SUBROUTINE DIAMOND

#include "fgl.h"
#include "fdevice.h"

C   draw grass
       CALL COLOR(9)
       CALL ARCFI (0, 0, 375, 0, 900)

       CALL COLOR(8)

C   change thickness of lines
       CALL LINEWI(2)

C   fences
       CALL ARCI(0, 0, 375, 0, 900)
       CALL ARCI(0, 0, 150, 0, 900)

C   foul lines
       CALL MOVE2I(0, 0)
       CALL DRAW2I(0, 400)
       CALL MOVE2I(0, 0)
       CALL DRAW2I(400, 0)

C   restore thickness of lines
       CALL LINEWI(1)

C   pitcher's mound
       CALL CIRCFI(43, 43, 10)

C   first, second and third base
       CALL DRAWBASE(90, 0)
       CALL DRAWBASE(90, 90)
       CALL DRAWBASE(0, 90)

C   draw home plate
       CALL PMV2I(0, 0)
       CALL PDR2I(0, 3)
       CALL PDR2I(3, 6)
       CALL PDR2I(6, 3)
       CALL PDR2I(3, 0)
       CALL PCLOS

C   draw scoreboard
       CALL CMOV2I (100, 400)
       CALL CHARST ('New York 3', 10)
       CALL CMOV2I (100, 385)
       CALL CHARST ('Boston   2', 10)

       RETURN
       END

C  draw_base has not changed since the last example.

       SUBROUTINE DRAWBASE(X, Y)

       INTEGER*2 X,Y

#include "fgl.h"
#include "fdevice.h"

       CALL PMV2I(X, Y)
       CALL RPDR2I(5, 0)
       CALL RPDR2I(0, 5)
       CALL RPDR2I(-5, 0)
       CALL PCLOS

       RETURN
       END

File: gifts.gl2/gl_tutor/src/workshop/double3.c
/*  double3.c  --  This program uses double buffering to draw a 	*/
/*		   dynamic image of a ball traveling over a field	*/
/*		   without flashing.  The motion is accomplished by	*/
/*		   redrawing the scene to the back buffer.  When an	*/
/*		   image is completed, the entire buffer is swapped	*/
/*		   to the fore, and its image is displayed.		*/
/*  calls introduced:  doublebuffer(), gconfig(), swapbuffers(),	*/
/*		       frontbuffer()					*/

/*  WARNING:  For machines with a minimum configuration of bitplanes,	*/
/*  there may not be enough bitplanes to run all the colors in this	*/
/*  and subsequent double buffered programs.  A minimum of 12		*/
/*  bitplanes are required for this program.  For subsequent programs,	*/
/*  a minimum of 16 bitplanes will be required.				*/

#include "gl.h"
#include "device.h"

/*  In the window manager, double buffered programs MUST swapbuffers,	*/
/*  even when idling.  The window manager waits for all active double	*/
/*  double buffered programs to issue a swapbuffers() call before	*/
/*  actually swapping buffers.  If a program idles without issuing	*/
/*  a swapbuffers() call, all other double buffered programs will	*/
/*  wait for it.							*/

main () {
    initialize ();
    drawimage ();
    while (TRUE) {	/*  loop forever.  swapbuffers while looping.	*/
	while (!qtest())/*  wait for input     				*/
	    swapbuffers();
	processinput(); /*  process input                               */
    }
}

/*  To initiate double buffering, the doublebuffer() call is made.	*/
/*  The display mode change to double buffering does not take place	*/
/*  until the gconfig() call is made.					*/

initialize () {

    prefsize (450, 450);
    winopen ("diamond");

    doublebuffer();	/* display mode to become double buffer */
    gconfig();		/* display mode change takes effect	*/

    qdevice( REDRAW );

    mapcolor (8, 240, 240, 240);/*  make color of arcs, lines	*/
    mapcolor (9, 0, 175, 0);	/*  make color of grass		*/
    mapcolor (10, 240, 150, 0);	/*  make color of ball		*/
}

/*  Process input from the window manager                               */

processinput() {
    short val;
    int dev;

    while (qtest()) {               /* while input on queue.            */
	dev = qread(&val);          /* read input device data.          */
	switch (dev) { 
	case REDRAW:                /* window manager asking for redraw.*/
	    reshapeviewport();      /* This shapes the window.          */
	    drawimage();            /* This redraws the image.          */
	    break;
	default:
	    break;
	}
    }
}

/*  drawimage() creates motion which doesn't flash.  The for loop	*/
/*  changes the value of i, which in turn, changes the position of	*/
/*  the ball.  Each iteration through the loop, the field is redrawn	*/
/*  and the ball is redrawn in a new position.				*/

drawimage () {
    int i;

    for (i = 0; i < 300; i = i + 3) {
	color (BLACK);		/*  clear away old field		*/
	clear ();
	diamond ();		/*  draw new field			*/
	color (10);
	circfs (i, i, 3);	/*  draw ball at position (x, y)	*/
	swapbuffers();
    }
	/*  At the end of the loop, the ball will stop moving.  However,*/
	/*  the two buffers will have the ball in different positions. 	*/
	/*  You want to be certain that the ball is in the same		*/
	/*  position in both buffers.  Otherwise, when at rest, the	*/
	/*  program will swap between the two buffers, and the ball	*/
	/*  will appear to bounce back and forth between the disparate	*/
	/*  positions in the two buffers.  To draw the same scene	*/
	/*  into both buffers, use frontbuffer(TRUE).  Then draw	*/
	/*  the scene--that draws it into both front and back buffers.	*/
	/*  frontbuffer(FALSE) restores the default state, where only	*/
	/*  the back buffer is drawn into.				*/
    frontbuffer(TRUE);	/*  draw final resting place of ball to		*/
    color(BLACK);	/*  both buffers				*/
    clear();
    diamond();
    color (10);
    circfs (i, i, 3);	/*  draw ball at position (x, y)	*/
    frontbuffer(FALSE);
}

/*  diamond() has not changed since the last example.		*/

diamond () {
    color (9);		/*  make color 9 current color	*/
    arcfs (0, 0, 375, 0, 900);	/*  grass		*/
    color (8);		/*  make color 8 current color	*/
    linewidth(2);	/*  change thickness of lines	*/
    arcs (0, 0, 375, 0, 900);	/*  fences		*/
    arcs (0, 0, 150, 0, 900);	/*  infield		*/

    move2s (0, 0);		/*  foul lines		*/
    draw2s (0, 400);
    move2s (0, 0);
    draw2s (400, 0);
    linewidth(1);	/*  restore thickness of lines	*/

    circfs (43, 43, 10);	/*  pitcher's mound	*/

    draw_base (90, 0);		/*  first, second and third bases	*/
    draw_base (90, 90);
    draw_base (0, 90);

    pmv2s (0, 0);		/*  draw home plate	*/
    pdr2s (0, 3);
    pdr2s (3, 6);
    pdr2s (6, 3);
    pdr2s (3, 0);
    pclos ();

    cmov2i (100, 400);	/*  position to draw scoreboard	*/
    charstr ("New York 3");
    cmov2i (100, 385);	
    charstr ("Boston   2");
}

/*  draw_base has not changed since the last example.			*/

draw_base (x, y)
short   x,
        y;
{
    pmv2s (x, y);
    rpdr2s (5, 0);
    rpdr2s (0, 5);
    rpdr2s (-5, 0);
    pclos ();
}
